File: Swift/max-area-of-island.swift
class Solution {
    func maxAreaOfIsland(_ grid: [[Int]]) -> Int {
        var array = grid
        var maxArea = 0

        for i in 0..<grid.count {
            for j in 0..<grid[0].count {
                if grid[i][j] == 1 {
                    maxArea = max(maxArea, dfs(&array, i: i, j: j))
                }
            }
        }

        return maxArea
    }

    private func dfs(_ grid:inout [[Int]], i: Int, j: Int) -> Int {
        if i < 0 || i >= grid.count || j < 0 || j >= grid[0].count || grid[i][j] == 0 {
            return 0
        }

        var temMaxArea = 1
        grid[i][j] = 0
        temMaxArea = temMaxArea + dfs(&grid, i: i-1, j: j) + dfs(&grid, i: i+1, j: j) + dfs(&grid, i: i, j: j-1) + dfs(&grid, i: i, j: j+1)

        return temMaxArea
    }
}

File: Python/max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        maxArea = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    maxArea = max(maxArea, self.dfs(grid, i, j))

        return maxArea

    def dfs(self, grid, i, j):
        if i < 0 or i >= len(grid) or j < 0 or j >= len(grid[0]) or grid[i][j] == 0:
            return 0

        temMaxArea = 1
        grid[i][j] = 0
        temMaxArea = temMaxArea + self.dfs(grid, i - 1, j) + self.dfs(grid, i + 1, j) + self.dfs(grid, i, j - 1) + self.dfs(grid, i, j + 1)
        return temMaxArea
